### PR TITLE
Remove continue point from Garbage Collector

### LINKED ISSUE #1351: Add test to GarbageCollectorIT for continue point or remove continue point
Accumulo garbage collection will read a batch of deletes into memory and process them.  When reading the next batch it uses the last path it saw as a continue point.  Need to ensure the continue point code covers all data.

Alternatively, instead of testing the continue point code it could be removed by refactoring the GC code as follows so that continue is not needed.  The code below keeps the iterator for the entire cycle instead of calling continue.

```java
Iterator<String> candidates = getContext().getAmple().getGcCandidates(level);

while(candidates.hasNext()){
     List<String> candidatesBatch =  readCandidatesThatFitInMemory(candidates);
    
    collectBatch(candidatesBatch);
}
```

### PR DESCRIPTION
**Pull request by jmark99**
Updated Garbage Collection code to no longer use a continue point when processing deletion candidates. The GC  now uses an iterator that lasts during the lifetime of a GC cycle.

The GarbageCollectionTest was updated to work with the update, as was the GC integration test.

Closes #1351

### CODE REVIEW

**Review by milleruntime — COMMENTED** [review]
I think these changes look good but I had to spend a good chunk of time learning how the GC currently uses the continue point. I would like to look at the changes closer and do some testing  but I would be OK if you merged before I get the chance.

**Comment by jmark99** [comment]
Thanks for taking a look @milleruntime. Looking more closely is a good idea since we definitely want the GC to work as expected. I've run the unit and IT tests as well as the sunny profile. Additionally, I ran the GC test in the accumulo-testing repo, although I'd like to run it for a longer period of time. I may wait and see if @keith-turner would like to take a look before merging since he had created the initial ticket.

File: server/gc/src/test/java/org/apache/accumulo/gc/GarbageCollectionTest.java
Comment (milleruntime) [inline server/gc/src/test/java/org/apache/accumulo/gc/GarbageCollectionTest.java:71]: Extra character in comment.
```suggestion
      // due to that causing a ConcurrentModificationException.
```

File: server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionEnvironment.java
Comment (milleruntime) [inline server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionEnvironment.java:45]: I wonder if this method should be split up. Previously we had one method to gather the candidates and then another that processed them.

File: server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java
Comment (milleruntime) [inline server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java:294]:
```suggestion
   * Given a sub-list of possible deletion candidates, process and remove valid deletion
```

File: server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java
Comment (milleruntime) [inline server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java:309]: The previous `collect()` method was confusing but it was the only public method of the `GarbageCollectionAlgorithm` so it was at least easier to follow from the one entry point. Since you are refactoring the methods, it might be a good opportunity to clean it up and make it easier to follow. The GC was already confusing but the way you split up the methods, is more confusing to me. It looks like you have the correct business logic between the methods but I think it could be organized better.

It seems the simplest solution would be to keep the same methods but just drop the continue point. Was there a reason you had to create new methods? I saw your comment in the tests about ConcurrentModificationException but I am not seeing where that could happen.

File: server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionEnvironment.java
Comment (jmark99) [reply server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionEnvironment.java:45]: Given that the processCandidates method is only about three lines, not counting the loop, I'm not sure it is worth it. I could re-create a method, getCandidates, that would be a one-liner returning the required Iterator and then gather the candidates and process them separately.

In the current code using a continue point, the getCandidates method  does get potential candidates and then returns to the collect method to do the processing. With the updated code using the iterator there is less of a separation between those two actions. Passing the iterator around between classes could become confusing as well. I had tried to do something similar to what you are suggesting but I was having issues with the iterator not keeping track of the candidates properly after being passed between multiple methods and classes.  Most likely coder error, but I eventually caved and went back to the current form that you see now.  I could try again if you feel strongly about it.

File: server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java
Comment (jmark99) [reply server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java:309]: I'll take another look, but primarily I was following @keith-turner format suggestion in the inital ticket. Since the call to Ample takes place in the classes implementing the GarbageCollectionEnvironment code, the few short lines seemed to belong in a single method. Not sure if better naming of methods would clear things up. 

Realistically the call to collect could be dropped altogether and just be replaced by a call to processCandidates (since that is all it calls anyway). I just kept it so the existing GC initial call would not be changed.  Then the GCA could have a  public method called' processCandidates' (or maybe even processDeletionCandidates). That seems clearer to me than using  'collect'. Then the the collectBatch call does what is implies. It collects a batch of potential deletion candidates and processes them. Maybe even get away from the 'collect' naming altogether and call it 'processBatch' or something similar.

File: server/gc/src/test/java/org/apache/accumulo/gc/GarbageCollectionTest.java
Comment (jmark99) [reply server/gc/src/test/java/org/apache/accumulo/gc/GarbageCollectionTest.java]: Any idea why making the two type changes you suggested causes the fastbuild to fail with a formatting error?

File: server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java
Comment (keith-turner) [reply server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java:309]: I suspect the ConcurrentModificationException occurs in the test.  I would suggest refactoring the test GCE so that this does not happen rather than the GCA.  The GarbageCollectionAlgorithm was designed to pull the logic of collection into a single place where it could be unit tested.  This change pushes some logic of collection into the GarbCollEnv, resulting in a need to change logic in the test which means the full algorithm is not being tested.

Comment (keith-turner) [reply server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java:309]: I suspect this is what was causing the CME in the unit test https://github.com/apache/accumulo/blob/146b1d3d0dc2ade6f1888c9905d83e37d98aedbe/server/gc/src/test/java/org/apache/accumulo/gc/GarbageCollectionTest.java#L81

File: server/gc/src/test/java/org/apache/accumulo/gc/GarbageCollectionTest.java
Comment (milleruntime) [reply server/gc/src/test/java/org/apache/accumulo/gc/GarbageCollectionTest.java]: White space maybe? I know the fastbuild checks if you built before pushing. Did you build locally after committing the suggestions? I almost always have to due to imports or formatting.

**Comment by keith-turner** [comment]
@jmark99 I just took an initial look at this.  I plan to take a more in-depth look later this afternoon.

File: server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionEnvironment.java
Comment (milleruntime) [reply server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionEnvironment.java:45]: My concern is creating more complexity in code that is already overly complex. If we are going to change the interface, lets design it in a way that reduces complexity. It appears that the GC interfaces were done so users could potentially write their own but I don't think that ever happened. This then makes me think that if we want to make the GC pluggable, lets do it right (either as a part of the SPI or API). But this seems like much too daunting of a task and I don't think there is much for users to gain by writing their own. That brings me to the idea that we should at least refactor the GC to make it more readable and maintainable.

Unfortunately, I don't have a suggestion on how that could be done for just these few methods. I would have to spend a lot more time looking at the GC.

File: server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java
Comment (jmark99) [reply server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java:309]: @keith-turner that was exactly the cause.

**Comment by EdColeman** [comment]
@mjwall might have additional insight if this overlaps with the issue that he has been tracking down.

File: server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java
Comment (keith-turner) [reply server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java:308]: Still have not looked in depth, so this may not be workable.  Thinking one possible way to avoid the CME would be to make the getCandidates() method in TestGCE return a snapshot.  Something like :

```java
   Iterator<String> getCandidates() {
        return List.copyOf(candidates).iterator();
   }
```

**Comment by jmark99** [comment]
@milleruntime , @keith-turner I did some re-factoring based upon @keith-turner suggestion concerning the getCandidates method. It now returns an interator. With that change the flow now more closely resembles the original flow of the code but without the continue point.

File: server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java
Comment (keith-turner) [inline server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java:306]:
```suggestion
  private void collectBatch(GarbageCollectionEnvironment gce, List<String> currentBatch)
```

File: test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorIT.java
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorIT.java:95]: Do you know if this size caused multiple batches to be processed in the IT?  If not could make it smaller.

**Review by keith-turner — APPROVED** [review]
@jmark99  the latest changes look great.  Made one small comment about making a method private, if that is possible.

File: test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorIT.java
Comment (jmark99) [reply test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorIT.java:95]: @keith-turner, yes. That was the reasoning behind the modification. With the gcLotsOfCandidatesIT test it has to process a lot of batches.

File: server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java
Comment (jmark99) [reply server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java]: Done. Thanks for catching this. It had been public in the prior form of the PR when it was being called from another class. Missed it when refactoring.

File: server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java
Comment (milleruntime) [inline server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java:306]: Could call this method `deleteBatch` since it doesn't do any collecting. That makes more sense to me since you already have the top level `collect()` method for the overall garbage collection.

**Review by milleruntime — APPROVED** [review]
Changes look better, thanks. Only other thing I had was the name of one method.

**Comment by mjwall** [comment]
I can look at this tonight if it is still open

File: server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java
Comment (jmark99) [reply server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java]: I made this change.

**Comment by jmark99** [comment]
@mjwall I will wait until tomorrow to merge if you wish to ltake a look tonight.

**Comment by mjwall** [comment]
This looks good to me @jmark99 .  It is unrelated to what I am tracking in #1916.  First I have looked at Ample, cool abstraction.  Hope it is enough &lt;pun intended&gt;

File: server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java
Comment (ctubbsii) [inline server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java:275]: This method doesn't make sense anymore. Since it's just returning an iterator and not actually doing all the fetching, tracing this operation isn't useful. The trace should be moved to wrap the retrieval of a batch.

File: server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java
Comment (ctubbsii) [inline server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java:298]: This line is what the `try (TraceScope candidatesSpan = Trace.startSpan("getCandidates")) {` should wrap.

**Review by ctubbsii — COMMENTED** [review]
I'm a bit late to this PR, but overall it looks great. I had only one comment about how we're tracing the candidate retrieval. Also I saw ConcurrentDeleteTableIT time out after this was committed, but I haven't had a chance to run it again to see if it's a fluke. Occasional timeouts are normal, so it might not be an issue at all.

File: server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java
Comment (jmark99) [reply server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java:298]: @ctubbsii, I created a separate PR to address this as the issue was already closed and merged. See #2216.

I'm hoping the IT timeout was a fluke. Not sure if it is included in the Sunny tests, but I have run them several times with no timeout.

**Pull request merged**
